start analysis vc

TOMalley104 opened the pull request:
Figuring out what to display has been ROUGH. The pie chart for songs by key is cool (or at least will be if we ever pull more than 100 songs) and I genuinely like the valence by key line chart. Statistics, man 😒

orieenav commented on spotifyMini/Models/Analysis.swift:
```diff
@@ -8,9 +8,70 @@
+
+    let name: String
+
+    class var keyTitles: [String] {
```
here, you could do static let keyTitles = ["C", "C#"... so that it's not a computed property with values that don't change

brightredchilli replied:
static let keyTitles = ["C", ...

orieenav commented on spotifyMini/Models/Analysis.swift:
```diff
@@ -8,9 +8,70 @@
+    var averageValenceByKey: [Float]? {
+        if let analyses = self.trackAnalyses {
+            var averageValues = [Float]()
+            for i in 0..<12 {
```
hm, where does this number come from?

orieenav commented on spotifyMini/Models/Analysis.swift:
```diff
@@ -8,9 +8,70 @@
+            var averageValues = [Float]()
+            for i in 0..<12 {
+                let analysesForThisKey = analyses.filter { $0.key == i }
+                let values = analysesForThisKey.map { $0.valence }
```
valences would make it clearer here

brightredchilli replied:
valences would be better here

brightredchilli commented on spotifyMini/Models/Analysis.swift:
```diff
@@ -8,9 +8,70 @@
+            for i in 0..<12 {
+                let analysesForThisKey = analyses.filter { $0.key == i }
+                let values = analysesForThisKey.map { $0.valence }
+                let totalValue = values.reduce(0) { $0 + $1 }
```
a little weird, but reduce(0, combine:+) also works

brightredchilli commented on spotifyMini/Models/Analysis.swift:
```diff
@@ -8,9 +8,70 @@
+    // MARK: Fetching
+
+    func fetchAnalyses() {
+        let ids = self.tracks.map { $0.identifier } as [String]
```
I think you don't need the `as [String]` part, would you give that a shot?

TOMalley104 replied:
"Cannot convert value of type '[String!]' to expected argument type '[String]'"

brightredchilli commented on spotifyMini/Models/Analysis.swift:
```diff
@@ -8,9 +8,70 @@
+        let ids = self.tracks.map { $0.identifier } as [String]
+        self.spotify.fetchAnalysis(forTrackIDs: ids) { result in
+            self.trackAnalyses = result.value
+            var userInfo: [String:AnyObject]?
```
userInfo is never instantiated here.

TOMalley104 replied:
as of "PR comments" commit, its instantiated within the `if let` that checks for an error

brightredchilli commented on spotifyMini/Utilities/Spotify.swift:
```diff
@@ -168,12 +168,26 @@ class Spotify {
             let task = NSURLSession.sharedSession().dataTaskWithRequest(request, completionHandler: { data, response, error in
                 if error != nil {
-                    completion(.Failure(SpotifyError.RequestFailed))
+                    Qu.Main {
```
So, the repetition of Qu.Main in several different places in the function is probably a good indication that something can be improved here. What you want is to do all of your parsing in the completionHandler, and have only one call to Qu.Main in the project. By the time you get into Qu.Main, you should already have enough information to decide between success or failure.

Two options: Either a)

```
NSURLSession.sharedSession().dataTaskWithRequest(request, completionHandler: { data, response, error in
  let results = .... // some work

  Qu.Main {
    if results != nil {
      completion() // success
    } else {
      completion(.SpotifyError(...)) // failure
    }
  }

}
```

brightredchilli commented on spotifyMini/ViewControllers/Analysis/AnalysisViewController.swift:
```diff
@@ -0,0 +1,158 @@
+        super.viewDidLoad()
+        let genre = "pop"
+        self.spotify.fetchRecommendedTracks(forGenre: genre) { result in
+            if let error = result.error as? SpotifyError {
```
will fetchRecommendedTracks ever return anything other than spotify error? if it doesn't then you can make the callback block return SpotifyError as opposed to ErrorType

TOMalley104 replied:
That happens because I'm using Intrepid's Result enum, who's `.Failure` case's associated value is `ErrorType`. Do you know of another way around this besides changing the pod itself?

brightredchilli commented on spotifyMini/ViewControllers/Analysis/AnalysisViewController.swift:
```diff
@@ -0,0 +1,158 @@
+
+        let lineDataSet = LineChartDataSet(yVals: lineDataEntries, label: "Avg. Valence by Key")
+        lineDataSet.axisDependency = .Left
+        lineDataSet.circleColors = colors ?? ChartColorTemplates.joyful()
```
joyful...nice

brightredchilli commented on spotifyMini/ViewControllers/Analysis/AnalysisViewController.swift:
```diff
@@ -0,0 +1,158 @@
+        }
+    }
+
+    func stopObservingAnalysisNotification() {
```
i think i missed this the last time, but you should also override deinit and call this method if you're targeting iOS8.0. On iOS9, upon dealloc, all notification observers are unregistered. not so on iOS 8 and below

TOMalley104 replied:
yup! we discussed it in the artist VC. This app is iOS 9 only so that code would be redundant (right?)

TOMalley104 commented:
added bar button on SearchVC to get to analysis VC, so you can reach all spots of the app without having to change what the storyboards initialViewController is. cleaned up all those `Qu` calls in Spotify.swift. Ready for re-review.
